ledger/streaming/publisher/src/grpc.rs

```diff
@@ -0,0 +1,230 @@
+}
+
+impl GrpcServerSink {
+    /// Instantiate a sink that publishes [SubscribeResponse]s over the
+    /// `LedgerUpdates` gRPC API.
+    pub fn new(logger: Logger) -> Self {
+        Self {
+            // Buffer a few responses.
+            publisher: Arc::new(Mutex::new(ExpiringPublisher::new(3))),
+            logger,
+        }
+    }
+
+    /// Consume a stream of `SubscribeResponse`.
+    /// The returned value is a `Stream` where the `Output` type is
+    /// `Result<()>`; it is executed entirely for its side effects, while
+    /// propagating errors back to the caller.
+    pub fn consume_protos<'a>(
+        &self,
+        stream: impl Stream<Item = Result<SubscribeResponse>> + 'a,
+    ) -> impl Stream<Item = Result<()>> + 'a {
+        let publisher = self.publisher.clone();
+        stream.and_then(move |data| {
+            let publisher = publisher.clone();
+            async move {
+                let mut publisher = publisher.lock().await;
+                publisher.publish(data).await;
+                Ok(())
+            }
+        })
+    }
+
+    /// Create a `LedgerUpdates` handler.
+    pub fn create_handler(&self) -> impl LedgerUpdates + Clone + Send + Sync + 'static {
+        let mut publisher = futures::executor::block_on(self.publisher.lock());
+        PublishHelper::new(publisher.subscribe())
+    }
+
+    /// Create a Service with a `LedgerUpdates` handler.
+    pub fn create_service(&self) -> grpcio::Service {
+        create_ledger_updates(self.create_handler())
+    }
+
+    /// Create a gRPC `Server` with a `LedgerUpdates` service using this
+    /// instance.
+    pub fn create_server(
+        &self,
+        uri: &impl ConnectionUri,
+        env: Arc<grpcio::Environment>,
+    ) -> grpcio::Result<grpcio::Server> {
+        grpcio::ServerBuilder::new(env)
+            .register_service(self.create_service())
+            .bind_using_uri(uri, self.logger.clone())
+            .build()
+    }
+}
+
+#[derive(Clone)]
+struct PublishHelper {
+    subscriber: Subscriber<SubscribeResponse>,
+}
+
+impl PublishHelper {
+    pub fn new(subscriber: Subscriber<SubscribeResponse>) -> Self {
+        Self { subscriber }
+    }
+}
+
+impl LedgerUpdates for PublishHelper {
+    fn subscribe(
+        &mut self,
+        ctx: grpcio::RpcContext,
+        req: SubscribeRequest,
+        sink: ServerStreamingSink<SubscribeResponse>,
+    ) {
+        let starting_height = req.starting_height;
+        let stream = self
+            .subscriber
+            .clone()
+            .skip_while(move |resp| {
+                let block_index = resp
+                    .get_result()
+                    .get_block()
+                    .get_v1()
+                    .get_block()
+                    .get_index();
+                futures::future::ready(block_index < starting_height)
+            })
+            .map(|resp| Ok((resp, grpcio::WriteFlags::default())));
+        let fut = stream.forward(sink).map(|_| ());
```

Comment:
Would protobuf be able to send an actual stream object across or must it send futures?

Reply:
I'm not sure what you mean